session/nontransactional.go

```diff
@@ -0,0 +1,551 @@
+			// `where x is null`
+			whereCondition = isNullCondition
+		} else {
+			// `where (x <= job.end) || (x is null)`
+			right := &driver.ValueExpr{}
+			right.Type = options.shardColumnType
+			right.Datum = job.end
+			leCondition := &ast.BinaryOperationExpr{
+				Op: opcode.LE,
+				L: &ast.ColumnNameExpr{
+					Name:  options.stmt.ShardColumn,
+					Refer: options.shardColumnRefer,
+				},
+				R: right,
+			}
+			whereCondition = &ast.BinaryOperationExpr{
+				Op: opcode.LogicOr,
+				L:  leCondition,
+				R:  isNullCondition,
+			}
+		}
+	} else {
+		// a normal between condition: `where x between start and end`
+		left := &driver.ValueExpr{}
+		left.Type = options.shardColumnType
+		left.Datum = job.start
+		right := &driver.ValueExpr{}
+		right.Type = options.shardColumnType
+		right.Datum = job.end
+		whereCondition = &ast.BetweenExpr{
+			Expr: &ast.ColumnNameExpr{
+				Name:  options.stmt.ShardColumn,
+				Refer: options.shardColumnRefer,
+			},
+			Left:  left,
+			Right: right,
+			Not:   false,
+		}
+	}
+
+	if options.originalCondition == nil {
+		options.stmt.DeleteStmt.Where = whereCondition
+	} else {
+		options.stmt.DeleteStmt.Where = &ast.BinaryOperationExpr{
+			Op: opcode.LogicAnd,
+			L:  whereCondition,
+			R:  options.originalCondition,
+		}
+	}
+	var sb strings.Builder
+	err := options.stmt.DeleteStmt.Restore(format.NewRestoreCtx(format.DefaultRestoreFlags|
+		format.RestoreNameBackQuotes|
+		format.RestoreSpacesAroundBinaryOperation|
+		format.RestoreBracketAroundBinaryOperation|
+		format.RestoreStringWithoutCharset, &sb))
+	if err != nil {
+		job.err = err
+		return ""
+	}
+	deleteSQL := sb.String()
+
+	if dryRun {
+		return deleteSQL
+	}
+
+	job.sql = deleteSQL
+	options.stmt.DeleteStmt.SetText(nil, fmt.Sprintf("/* job %v/%v */ %s", job.jobID, totalJobCount, deleteSQL))
+	rs, err := se.ExecuteStmt(ctx, options.stmt.DeleteStmt)
+
+	// collect errors
+	failpoint.Inject("splitDeleteError", func(_ failpoint.Value) {
+		err = errors.New("injected split delete error")
+	})
+	if err != nil {
+		errStr := fmt.Sprintf("Non-transactional delete SQL failed, sql: %s, error: %s, jobID: %d, jobSize: %d. ",
+			deleteSQL, err.Error(), job.jobID, job.jobSize)
+		logutil.Logger(ctx).Error(errStr)
+		job.err = err
+	} else {
+		logutil.Logger(ctx).Info("Non-transactional delete SQL finished successfully", zap.Int("jobID", job.jobID),
+			zap.Int("jobSize", job.jobSize), zap.String("deleteSQL", deleteSQL))
+	}
+	if rs != nil {
+		rs.Close()
+	}
+	return ""
+}
+
+func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDeleteStmt, se Session,
+	selectSQL string, shardColumnInfo *model.ColumnInfo) ([]job, error) {
+	logutil.Logger(ctx).Info("Non-transactional delete, select SQL", zap.String("selectSQL", selectSQL))
+	var shardColumnCollate string
+	if shardColumnInfo != nil {
+		shardColumnCollate = shardColumnInfo.Collate
+	} else {
+		shardColumnCollate = ""
+	}
+
+	rss, err := se.Execute(context.TODO(), selectSQL)
```

Comment:
Just for learning: why here use `context.TODO()` other than input parameter `ctx` ?

Reply:
Ah we should use `ctx`.